capsules/src/kv_store.rs

```diff
@@ -0,0 +1,434 @@
+        flash: &'a F,
+        data_buffer: &'static mut F::Page,
+        region_offset: usize,
+    ) -> TickFSFlastCtrl<'a, F> {
+        Self {
+            flash,
+            data_buffer: TakeCell::new(data_buffer),
+            state: Cell::new(State::None),
+            region_offset,
+        }
+    }
+}
+
+impl<'a, F: Flash> tickfs::flash_controller::FlashController for TickFSFlastCtrl<'a, F> {
+    fn read_region(
+        &self,
+        region_number: usize,
+        _offset: usize,
+        buf: &mut [u8],
+    ) -> Result<(), tickfs::error_codes::ErrorCode> {
+        match self.state.get() {
+            State::ReadComplete(reg) => {
+                if reg as usize == region_number {
+                    // We already have read the data.
+                    let data_buf = self.data_buffer.take().unwrap();
+                    for (i, d) in data_buf.as_mut().iter().enumerate() {
+                        buf[i] = *d;
+                    }
+                    self.data_buffer.replace(data_buf);
+                    self.state.set(State::None);
+
+                    return Ok(());
+                }
+
+                if self
+                    .flash
+                    .read_page(
+                        self.region_offset + region_number,
+                        self.data_buffer.take().unwrap(),
+                    )
+                    .is_err()
+                {
+                    return Err(tickfs::error_codes::ErrorCode::ReadFail);
+                }
+            }
+            _ => {
+                if self
+                    .flash
+                    .read_page(
+                        self.region_offset + region_number,
+                        self.data_buffer.take().unwrap(),
+                    )
+                    .is_err()
+                {
+                    return Err(tickfs::error_codes::ErrorCode::ReadFail);
+                }
+            }
+        }
+
+        Err(tickfs::error_codes::ErrorCode::ReadNotReady(region_number))
+    }
+
+    fn write_region(
+        &self,
+        address: usize,
+        buf: &[u8],
+    ) -> Result<(), tickfs::error_codes::ErrorCode> {
+        let data_buf = self.data_buffer.take().unwrap();
+
+        for (i, d) in buf.iter().enumerate() {
+            data_buf.as_mut()[i] = *d;
+        }
+
+        if self
+            .flash
+            .write_page((0x20040000 + address) / 1024, data_buf)
+            .is_err()
+        {
+            return Err(tickfs::error_codes::ErrorCode::WriteFail);
+        }
+
+        Err(tickfs::error_codes::ErrorCode::WriteNotReady(address))
+    }
+
+    fn erase_region(&self, region_number: usize) -> Result<(), tickfs::error_codes::ErrorCode> {
+        self.flash.erase_page(self.region_offset + region_number);
+
+        Err(tickfs::error_codes::ErrorCode::EraseNotReady(region_number))
+    }
+}
+
+pub struct KVStoreDriver<'a, F: Flash + 'static> {
+    tickfs: TickFS<'a, TickFSFlastCtrl<'a, F>, SipHasher>,
```

Comment:
this setup, where the userspace driver owns the filesystem, suggests that the file system will only be for use by applications. I think this is fine as an initial approach, but it would ultimately be nice for capsules to be able to write to the file system. Do you think anything about the design of TickFS precludes virtualization within the kernel?

Reply:
Yeah, my current focus has only been exposing it to userspace applications.

I can't think of a reason we can't virtualise it, but I haven't thought about it much to be honest.

Reply:
Sounds good. Given that TickFS does not implement a HIL-like interface, it seems fine to just punt this problem to the future.